File: tests/nsstring-path-properties.m
#import <Foundation/Foundation.h>
#include <stdlib.h>
#include <string.h>

static void expect(BOOL condition, NSString *message)
{
    if (!condition)
    {
        NSLog(@"FAIL: %@", message);
        exit(1);
    }
}

int main(void)
{
    @autoreleasepool
    {
        NSString *path = @"/tmp/darling/report.tar.gz";
        expect([path.lastPathComponent isEqualToString:@"report.tar.gz"], @"lastPathComponent");
        expect([path.pathExtension isEqualToString:@"gz"], @"pathExtension");
        expect([path.stringByDeletingPathExtension isEqualToString:@"/tmp/darling/report.tar"], @"stringByDeletingPathExtension");
        expect([path.stringByDeletingLastPathComponent isEqualToString:@"/tmp/darling"], @"stringByDeletingLastPathComponent");
        expect([path.pathComponents isEqualToArray:(@[ @"/", @"tmp", @"darling", @"report.tar.gz" ])], @"pathComponents");
        expect(path.absolutePath && path.isAbsolutePath, @"absolutePath");
        expect(!@"relative/path".absolutePath, @"relative path is not absolute");
        expect([@"".lastPathComponent isEqualToString:@""], @"empty lastPathComponent");
        expect([@"/a/./b/../c".stringByStandardizingPath isEqualToString:@"/a/c"], @"stringByStandardizingPath");

        NSString *home = NSHomeDirectory();
        NSString *expanded = @"~/notes.txt".stringByExpandingTildeInPath;
        expect([expanded isEqualToString:[home stringByAppendingPathComponent:@"notes.txt"]], @"stringByExpandingTildeInPath");
        if (![home isEqualToString:@"/"])
        {
            expect([expanded.stringByAbbreviatingWithTildeInPath isEqualToString:@"~/notes.txt"], @"stringByAbbreviatingWithTildeInPath");
        }
        expect([@"/".stringByResolvingSymlinksInPath isEqualToString:@"/"], @"stringByResolvingSymlinksInPath");
        expect(strcmp(@"/tmp/x".fileSystemRepresentation, "/tmp/x") == 0, @"fileSystemRepresentation");
    }
    NSLog(@"PASS: NSString path properties");
    return 0;
}

File: tests/nsstring-path-properties.swift
import Foundation

func expect(_ condition: Bool, _ message: String) {
    if !condition {
        print("FAIL: \(message)")
        exit(1)
    }
}

// As on macOS, the NSString path accessors import as non-optional properties.
let path = "/tmp/darling/report.tar.gz" as NSString
let last: String = path.lastPathComponent
let directory: String = path.deletingLastPathComponent
let components: [String] = path.pathComponents
let absolute: Bool = path.isAbsolutePath
let representation: UnsafePointer<CChar> = path.fileSystemRepresentation
let appended: String? = path.appendingPathExtension("bak")
expect(last == "report.tar.gz", "lastPathComponent")
expect(directory == "/tmp/darling", "deletingLastPathComponent")
expect(path.pathExtension == "gz", "pathExtension")
expect(path.deletingPathExtension == "/tmp/darling/report.tar", "deletingPathExtension")
expect(components == ["/", "tmp", "darling", "report.tar.gz"], "pathComponents")
expect(absolute, "isAbsolutePath")
expect(String(cString: representation) == "/tmp/darling/report.tar.gz", "fileSystemRepresentation")
expect(appended == "/tmp/darling/report.tar.gz.bak", "appendingPathExtension")
expect(("/a/./b/../c" as NSString).standardizingPath == "/a/c", "standardizingPath")
print("PASS: NSString path properties")
